File: LC2130.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        t1 = t2 = head
        st = []

        while t2 and t2.next:
            st.append(t1)
            t1 = t1.next
            t2 = t2.next.next
        
        m = 0

        while t1:
            m = max(m, t1.val + st.pop().val)
            t1 = t1.next
        
        return m

File: LC2130.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        stack <ListNode*> st;
        ListNode *t1 = head, *t2 = head;

        while(t2 and t2->next){
            st.push(t1);
            t1 = t1->next;
            t2 = t2->next->next;
        }

        int m = 0;
        while(t1){
            m = max(m, t1->val + st.top()->val);
            t1 = t1->next;
            st.pop();
        }

        return m;
    }
};
